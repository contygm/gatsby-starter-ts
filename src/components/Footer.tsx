import * as React from 'react';
import { Link } from 'gatsby';
import { StaticImage } from 'gatsby-plugin-image';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faStar, faMugHot, faGhost, faAngleUp } from '@fortawesome/free-solid-svg-icons';
import { IconProp } from '@fortawesome/fontawesome-svg-core';

/**
 * @description A icon link using FontAwesome icons with
 * footer styles
 * @param faIcon - IconProp, faicon name
 */
 const FaIconLink = (props: { faIcon: IconProp }) => {
    return (
        <li className="mx-2 is-inline">
            <Link
                className="has-text-info-light"
                to="/"
            >
                <FontAwesomeIcon
                    icon={props.faIcon}
                    size="lg"
                />
            </Link>
        </li>
    );
};

/**
 * @description A text link with footer styles
 * @param pageUrl - string, destination url
 * @param label - string, link text
 */
const PageLink = (props: { pageUrl: string; label: string }) => {
    return (
        <li className="is-inline is-inline-desktop is-inline-tablet mx-3">
            <Link
                className="has-text-info-light"
                to={props.pageUrl}
            >
                {props.label}
            </Link>
        </li>
    );
};

/**
 * @description A link dropdown that routes to the legal pages: Copyright, 
 * Cookie Policy, Terms and Conditions, Disclaimer. Dropdown actually drops UP
 */
const DropDownLinkBtn = () => {
    return (
        <li className="dropdown is-hoverable is-up">
            <div className="dropdown-trigge">
                <Link 
                    className="has-text-info-light ml-3"
                    aria-haspopup="true"
                    aria-controls="footer-dropdown-menu"
                    to="/legal"
                >
                    Legal
                    <FontAwesomeIcon
                        icon={faAngleUp}
                        size="sm"
                        className="px-2"
                    />
                </Link>
            </div>
            <div className="dropdown-menu" id="footer-dropdown-menu" role="menu">
                <div className="dropdown-content">
                    <div className="dropdown-item">
                        <Link className="" to="/legal/copyright">Copyright</Link>
                    </div>
                    <div className="dropdown-item">
                        <Link className="" to="/legal/cookie-policy">Cookie Policy</Link>
                    </div>
                    <div className="dropdown-item">
                        <Link className="" to="/legal/terms-and-conditions">Terms and Conditions</Link>
                    </div>
                    <div className="dropdown-item">
                        <Link className="" to="/legal/disclaimer">Disclaimer</Link>
                    </div>
                </div>
            </div>
        </li>
    );
}

/**
 * @description Site wide footer, mobile friendly
 * @param author - author of site, passed in from layour
 */
const Footer = (props: { author: string }) => {
    return (
        <footer className="footer">
            <div className="columns mb-1 is-vcentered has-text-centered-mobile">
                {/* Logo to the left */}
                <div className="column">
                    <Link
                        className=""
                        to="/"
                    >
                        <StaticImage
                            src="../images/icon.png"
                            alt="Logo"
                            placeholder="blurred"
                            className="footer-logo"
                            layout="constrained"
                            width={50}
                            height={50}
                        />
                    </Link>
                </div>
                {/* Navigation links centered */}
                <ul className="column has-text-centered">
                    {/* TODO: display: flex for div will get tthe links to site straight  */}
                    {/* TODO: need to account for mobile tho */}
                    {/* https://stackoverflow.com/questions/27418104/whats-the-difference-between-displayinline-flex-and-displayflex */}
                    <div >
                        <PageLink
                            label="Contact"
                            pageUrl="/"
                        />
                        <PageLink
                            label="About"
                            pageUrl="/"
                        />
                        <PageLink
                            label="Privacy"
                            pageUrl="/legal/privacy-policy"
                        />
                        {/* contians the rest of the 'legal' pages */}
                        <DropDownLinkBtn />
                    </div>
                    
                </ul>
                {/* Social media links to the left */}
                <div className="column">
                    <ul className="has-text-right has-text-centered-mobile">
                        <FaIconLink faIcon={faStar} />
                        <FaIconLink faIcon={faMugHot} />
                        <FaIconLink faIcon={faGhost} />
                    </ul>
                </div>
            </div>
            {/* Copy right section with border */}
            {/* TODO: use <time /> */}
            <div className="footer-copyright has-text-centered has-text-grey-light is-size-7">
                <p className="pt-3">
                    Â© {new Date().getFullYear()}, Built by {props.author}
                </p>
            </div>
        </footer>
    );
};

export default Footer;
