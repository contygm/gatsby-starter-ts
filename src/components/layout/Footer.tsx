import React from 'react';
import { Link } from 'gatsby';
import { StaticImage } from 'gatsby-plugin-image';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
    faStar,
    faMugHot,
    faGhost,
    faAngleUp
} from '@fortawesome/free-solid-svg-icons';
import { FaIconLink } from '../common/FaIconLink';
import { LEGAL_LINKS } from '../../utils/constants/legalDropdownLinks';

/**
 * A text link with footer styles
 * @param {string} pageUrl - string, destination url
 * @param {string} label - string, link text
 *
 * @memberof Footer
 */
const PageLink = (props: { pageUrl: string; label: string }) => {
    return (
        <li
            className="footer-pg-link-wrapper"
            role="listitem"
        >
            <Link
                className="footer-link-base"
                to={props.pageUrl}
                data-cy={`footer-link-${props.label}`}
            >
                {props.label}
            </Link>
        </li>
    );
};

/**
 * A link with dropdown styles that will go to a specific legal page
 * @param {string} label - string, link text
 * @param {string} slug - string, destination url
 *
 * @memberof Footer
 */
const DropUpLink = (props: { label: string; slug: string }) => {
    return (
        <div className="dropdown-item">
            <Link
                className="footer-drawer-link"
                to={`${props.slug}`}
                data-cy={`footer-link-${props.slug}`}
            >
                {props.label}
            </Link>
        </div>
    );
};

/**
 * A link dropdown that routes to the legal page index and the individual legal pages (hardcoded values):
 * - Copyright
 * - Cookie Policy
 * - Terms and Conditions
 * - Disclaimer.
 *
 * Dropdown actually drops UP.
 * @memberof Footer
 */
const DropUpBtn = (props: {
    links: {
        label: string;
        slug: string;
    }[]
}) => {
    return (
        <li
            className="footer-dropup"
            role="listitem"
        >
            <div className="dropdown-trigger">
                <Link
                    className="footer-dropup-menu"
                    aria-haspopup="true"
                    aria-controls="footer-dropdown-menu"
                    to="/legal"
                    data-cy="footer-link-legal"
                >
                    Legal
                    <FontAwesomeIcon
                        icon={faAngleUp}
                        size="sm"
                        className="px-2"
                    />
                </Link>
            </div>
            <div
                className="dropdown-menu"
                id="footer-dropdown-menu"
                role="menu"
            >
                <div className="dropdown-content">
                    {props.links.map((link) => (
                        <DropUpLink
                            key={link.label}
                            label={link.label}
                            slug={link.slug}
                        />
                    ))}
                </div>
            </div>
        </li>
    );
};

/**
 * Site-wide, mobile-friendly footer that includes:
 * - copyright with site author
 * - legal page links in a dropdown button
 * - site logo
 * - links to privacy, contact and about pages
 * - social media links
 * @param author - author of site, passed in from layout
 *
 * @category Components
 */
export const Footer = (props: { author: string }) => {
    return (
        <footer className="footer">
            <div className="footer-main-content-wrapper">
                {/* Logo to the left */}
                <div className="column">
                    <Link
                        className=""
                        to="/"
                        data-cy="footer-logo"
                    >
                        <figure className="footer-logo">
                            <StaticImage
                                src="../../images/bulma-logo-white.png"
                                alt="Logo"
                                placeholder="blurred"
                                layout="constrained"
                                width={128}
                            />
                        </figure>
                    </Link>
                </div>
                {/* Navigation links centered */}
                {/* NOTE: flex basis will keep dropdown btn in line with the links */}
                <ul
                    className="centered-col"
                    style={{ flexBasis: 'auto' }}
                    role="list"
                >
                    <div>
                        <PageLink
                            label="Contact"
                            pageUrl="/about#contact-us"
                        />
                        <PageLink
                            label="About"
                            pageUrl="/about"
                        />
                        <PageLink
                            label="Privacy"
                            pageUrl="/legal/privacy-policy"
                        />
                        {/* contains the rest of the 'legal' pages */}
                        <DropUpBtn links={LEGAL_LINKS}/>
                    </div>
                </ul>
                {/* Social media links to the left */}
                <div className="column">
                    <ul className="footer-social-media">
                        <FaIconLink faIcon={faStar} />
                        <FaIconLink faIcon={faMugHot} />
                        <FaIconLink faIcon={faGhost} />
                    </ul>
                </div>
            </div>
            {/* Copy right section with border */}
            <div className="footer-copyright">
                <p
                    className="pt-3"
                    data-testid="copyright"
                >
                    Â© <time>{new Date().getFullYear()}</time>, Built by{' '}
                    {props.author}
                </p>
            </div>
        </footer>
    );
};
