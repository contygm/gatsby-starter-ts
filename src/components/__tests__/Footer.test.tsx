import React from 'react';
import { render } from '@testing-library/react';
import {Footer} from '../Footer';

describe('Footer', () => {
    it('renders correctly', () => {
        const author = 'Author';

        const { asFragment, getByTestId } = render(<Footer author={author} />);
        expect(asFragment()).toMatchSnapshot();
        expect(getByTestId('copyright').textContent).toContain(author);
    });
});
