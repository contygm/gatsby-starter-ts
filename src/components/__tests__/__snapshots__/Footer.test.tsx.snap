// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer renders correctly 1`] = `
<DocumentFragment>
  <footer
    class="footer"
  >
    <div
      class="columns mb-1 is-vcentered has-text-centered-mobile"
    >
      <div
        class="column"
      >
        <a
          class=""
          data-cy="footer-logo"
          href="/"
        >
          <figure
            class="image footer-logo"
          >
            <img
              alt="Logo"
              layout="constrained"
              placeholder="blurred"
              src="../images/bulma-logo-white.png"
              width="128"
            />
          </figure>
        </a>
      </div>
      <ul
        class="column has-text-centered"
        role="list"
        style="flex-basis: auto;"
      >
        <div>
          <li
            class="is-inline-desktop is-inline-tablet m-4 footer-link"
            role="listitem"
          >
            <a
              class="footer-link"
              data-cy="footer-link-Contact"
              href="/about#contact-us"
            >
              Contact
            </a>
          </li>
          <li
            class="is-inline-desktop is-inline-tablet m-4 footer-link"
            role="listitem"
          >
            <a
              class="footer-link"
              data-cy="footer-link-About"
              href="/about"
            >
              About
            </a>
          </li>
          <li
            class="is-inline-desktop is-inline-tablet m-4 footer-link"
            role="listitem"
          >
            <a
              class="footer-link"
              data-cy="footer-link-Privacy"
              href="/legal/privacy-policy"
            >
              Privacy
            </a>
          </li>
          <li
            class="dropdown is-hoverable is-up"
            role="listitem"
          >
            <div
              class="dropdown-trigger"
            >
              <a
                aria-controls="footer-dropdown-menu"
                aria-haspopup="true"
                class="m-4 footer-link"
                data-cy="footer-link-legal"
                href="/legal"
              >
                Legal
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-angle-up fa-sm px-2"
                  data-icon="angle-up"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 448 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M201.4 137.4c12.5-12.5 32.8-12.5 45.3 0l160 160c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L224 205.3 86.6 342.6c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l160-160z"
                    fill="currentColor"
                  />
                </svg>
              </a>
            </div>
            <div
              class="dropdown-menu"
              id="footer-dropdown-menu"
              role="menu"
            >
              <div
                class="dropdown-content"
              >
                <div
                  class="dropdown-item"
                >
                  <a
                    class="footer-drawer-link"
                    data-cy="footer-link-copyright"
                    href="/legal/copyright"
                  >
                    Copyright
                  </a>
                </div>
                <div
                  class="dropdown-item"
                >
                  <a
                    class="footer-drawer-link"
                    data-cy="footer-link-cookie"
                    href="/legal/cookie-policy"
                  >
                    Cookie Policy
                  </a>
                </div>
                <div
                  class="dropdown-item"
                >
                  <a
                    class="footer-drawer-link"
                    data-cy="footer-link-terms"
                    href="/legal/terms-and-conditions"
                  >
                    Terms and Conditions
                  </a>
                </div>
                <div
                  class="dropdown-item"
                >
                  <a
                    class="footer-drawer-link"
                    data-cy="footer-link-disclaimer"
                    href="/legal/disclaimer"
                  >
                    Disclaimer
                  </a>
                </div>
              </div>
            </div>
          </li>
        </div>
      </ul>
      <div
        class="column"
      >
        <ul
          class="has-text-right has-text-centered-mobile footer-link"
        >
          <li
            class="faicon-link-container is-inline"
          >
            <a
              aria-label="social media link"
              class="faicon-link no-label"
              data-cy="faicon-link"
              href="/"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-star fa-xl icon"
                data-icon="star"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M316.9 18C311.6 7 300.4 0 288.1 0s-23.4 7-28.8 18L195 150.3 51.4 171.5c-12 1.8-22 10.2-25.7 21.7s-.7 24.2 7.9 32.7L137.8 329 113.2 474.7c-2 12 3 24.2 12.9 31.3s23 8 33.8 2.3l128.3-68.5 128.3 68.5c10.8 5.7 23.9 4.9 33.8-2.3s14.9-19.3 12.9-31.3L438.5 329 542.7 225.9c8.6-8.5 11.7-21.2 7.9-32.7s-13.7-19.9-25.7-21.7L381.2 150.3 316.9 18z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </li>
          <li
            class="faicon-link-container is-inline"
          >
            <a
              aria-label="social media link"
              class="faicon-link no-label"
              data-cy="faicon-link"
              href="/"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-mug-hot fa-xl icon"
                data-icon="mug-hot"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M88 0C74.7 0 64 10.7 64 24c0 38.9 23.4 59.4 39.1 73.1l1.1 1C120.5 112.3 128 119.9 128 136c0 13.3 10.7 24 24 24s24-10.7 24-24c0-38.9-23.4-59.4-39.1-73.1l-1.1-1C119.5 47.7 112 40.1 112 24c0-13.3-10.7-24-24-24zM32 192c-17.7 0-32 14.3-32 32V416c0 53 43 96 96 96H288c53 0 96-43 96-96h16c61.9 0 112-50.1 112-112s-50.1-112-112-112H352 32zm352 64h16c26.5 0 48 21.5 48 48s-21.5 48-48 48H384V256zM224 24c0-13.3-10.7-24-24-24s-24 10.7-24 24c0 38.9 23.4 59.4 39.1 73.1l1.1 1C232.5 112.3 240 119.9 240 136c0 13.3 10.7 24 24 24s24-10.7 24-24c0-38.9-23.4-59.4-39.1-73.1l-1.1-1C231.5 47.7 224 40.1 224 24z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </li>
          <li
            class="faicon-link-container is-inline"
          >
            <a
              aria-label="social media link"
              class="faicon-link no-label"
              data-cy="faicon-link"
              href="/"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-ghost fa-xl icon"
                data-icon="ghost"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 384 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M50.8 452.1L19.2 477.4c-2.1 1.7-4.7 2.6-7.4 2.6C5.3 480 0 474.7 0 468.2V192C0 86 86 0 192 0S384 86 384 192V468.2c0 6.5-5.3 11.8-11.8 11.8c-2.7 0-5.3-.9-7.4-2.6l-31.6-25.3c-3.3-2.7-7.5-4.1-11.8-4.1c-5.9 0-11.5 2.8-15 7.5l-37.6 50.1c-3 4-7.8 6.4-12.8 6.4s-9.8-2.4-12.8-6.4l-38.4-51.2c-3-4-7.8-6.4-12.8-6.4s-9.8 2.4-12.8 6.4l-38.4 51.2c-3 4-7.8 6.4-12.8 6.4s-9.8-2.4-12.8-6.4L77.6 455.5c-3.6-4.7-9.1-7.5-15-7.5c-4.3 0-8.4 1.5-11.7 4.1zM160 192c0-17.7-14.3-32-32-32s-32 14.3-32 32s14.3 32 32 32s32-14.3 32-32zm96 32c17.7 0 32-14.3 32-32s-14.3-32-32-32s-32 14.3-32 32s14.3 32 32 32z"
                  fill="currentColor"
                />
              </svg>
            </a>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="footer-copyright has-text-centered has-text-grey-light is-size-7"
    >
      <p
        class="pt-3"
        data-testid="copyright"
      >
        Â© 
        <time>
          2022
        </time>
        , Built by Author
      </p>
    </div>
  </footer>
</DocumentFragment>
`;
